packages/react/src/TextInput/TextInput.tsx

```diff
@@ -134,64 +151,142 @@ const TextInput = React.forwardRef<HTMLInputElement, TextInputProps>(
-          // @ts-expect-error it needs a non nullable ref
-          ref={inputRef}
+      <>
+        <TextInputWrapper
+          block={block}
+          className={wrapperClasses}
+          validationStatus={isValid}
+          contrast={contrast}
           disabled={disabled}
-          onFocus={handleInputFocus}
-          onBlur={handleInputBlur}
-          type={type}
-          aria-required={required}
-          aria-invalid={validationStatus === 'error' ? 'true' : undefined}
-          {...inputProps}
-          aria-describedby={inputDescribedBy}
-          data-component="input"
-        />
-        {loading && <VisuallyHidden id={loadingId}>{loaderText}</VisuallyHidden>}
-        <TextInputInnerVisualSlot
-          visualPosition="trailing"
-          showLoadingIndicator={showTrailingLoadingIndicator}
-          hasLoadingIndicator={typeof loading === 'boolean'}
-          id={trailingVisualId}
-          data-testid="text-input-trailing-visual"
+          monospace={monospace}
+          size={sizeProp}
+          width={widthProp}
+          minWidth={minWidthProp}
+          maxWidth={maxWidthProp}
+          variant={variantProp}
+          hasLeadingVisual={Boolean(LeadingVisual || showLeadingLoadingIndicator)}
+          hasTrailingVisual={Boolean(TrailingVisual || showTrailingLoadingIndicator)}
+          hasTrailingAction={Boolean(trailingAction)}
+          isInputFocused={isInputFocused}
+          onClick={focusInput}
+          aria-busy={Boolean(loading)}
         >
-          {typeof TrailingVisual !== 'string' && isValidElementType(TrailingVisual) ? (
-            <TrailingVisual />
-          ) : (
-            TrailingVisual
-          )}
-        </TextInputInnerVisualSlot>
-        {trailingAction}
-      </TextInputWrapper>
+          {IconComponent && <IconComponent className="TextInput-icon" />}
+          <TextInputInnerVisualSlot
+            visualPosition="leading"
+            showLoadingIndicator={showLeadingLoadingIndicator}
+            hasLoadingIndicator={typeof loading === 'boolean'}
+            id={leadingVisualId}
+          >
+            {typeof LeadingVisual !== 'string' && isValidElementType(LeadingVisual) ? <LeadingVisual /> : LeadingVisual}
+          </TextInputInnerVisualSlot>
+          <UnstyledTextInput
+            // @ts-expect-error it needs a non nullable ref
+            ref={inputRef}
+            disabled={disabled}
+            onFocus={handleInputFocus}
+            onBlur={handleInputBlur}
+            onChange={handleInputChange}
+            type={type}
+            aria-required={required}
+            aria-invalid={isValid === 'error' ? 'true' : undefined}
+            value={value}
+            defaultValue={defaultValue}
+            {...inputProps}
+            aria-describedby={
+              characterLimit
+                ? [characterCountStaticMessageId, inputDescribedBy].filter(Boolean).join(' ') || undefined
+                : inputDescribedBy
+            }
```

Comment:
This part changed in order to add the static message ID (same note as above for the `TextArea` component).